docs/articles/test-workflows-high-level-architecture.md

```diff
@@ -0,0 +1,166 @@
+This way, they will be acknowledged as free of side effects, and will be merged with next or previous step when it is possible.
+
+```yaml
+spec:
+  steps:
+  # container 1
+  - shell: echo 'Hello world 1'
+    pure: true
+  - shell: echo 'Hello world 2'
+  # container 2
+  - shell: echo 'Hello world 3'
+  - shell: echo 'Hello world 4'
+    pure: true
+  - shell: echo 'Hello world 5'
+```
+
+:::info
+
+If you wish to have that behavior by default for all steps, you can use system flag in the Test Workflow or [**Global Template**](./test-workflows-job-and-pod.md#global-template):
+
+```yaml
+spec:
+  system:
+    pureByDefault: true
+```
+
+:::
+
+![Pure Workflow Containers](./images/pure-workflow-containers.png)
+
+## Init Process
+
+To perform advanced orchestration, Test Workflows are using own [**Init Process**](https://en.wikipedia.org/wiki/Init).
+That process is an actual command run in the container, that is preparing the environment, and executing the actual command to run.
+
+Whenever you run container with some command, like `command: [ "curl", "http://testkube.io" ]`,
+we are actually doing something similar to `command: [ "/init", "curl", "http://testkube.io" ]`.
+The `/init` process is responsible for all the preparation, execution of actual command, and cleanup.
+
+## Fetching Image Metadata from the Container Registry
+
+To build the proper Pod and Init Process, we need to know the most important details of the image, like:
+* **Working Directory (`workingDir` / [`WORKDIR`](https://docs.docker.com/reference/dockerfile/#workdir)):** to change current working directory between operations in a single container
+* **Command (`command` / [`ENTRYPOINT`](https://docs.docker.com/reference/dockerfile/#entrypoint) and `args` / [`CMD`](https://docs.docker.com/reference/dockerfile/#cmd)):** to wrap with the Init Process
+* **User and Group (`securityContext.runAsUser/runAsGroup` / [`USER`](https://docs.docker.com/reference/dockerfile/#user)):** to configure the file system
+
+Kubernetes doesn't have a direct mechanism to get these metadata. To obtain it, we need to fetch the data from Container Registry when we need.
+
+### Private Container Registries
+
+When the image you are using is stored in the private registry, you may need to add [`pod.imagePullSecrets`](https://docs.testkube.io/articles/test-workflows-examples-basics#configuring-the-pod) to the specific Test Workflow or [**Global Template**](https://docs.testkube.io/articles/test-workflows-job-and-pod#global-template).
+
+### Avoid Fetching Image Metadata
+
+Companies may have a strict policy of not storing the credentials for the Container Registry in the Kubernetes' Secret, or disallow Secrets at all.
+
+When Testkube will not have access to the Container Registry, it may revoke running the Test Workflow with errors about not being able to access the private image metadata.
```

Comment:
```suggestion
When Testkube does not have access to the Container Registry, it may revoke running the Test Workflow with errors about not being able to access the private image metadata.
```